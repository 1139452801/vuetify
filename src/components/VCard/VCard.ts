// Styles
import '../../stylus/components/_cards.styl'

// Components
import { VPaper } from '../VPaper'

// Mixins
import Routable from '../../mixins/routable'

// Helpers
import { addPaperClasses } from '../VPaper/VPaper'
import { convertToUnit } from '../../util/helpers'
import { deprecate } from '../../util/console'
import mixins from '../../util/mixins'

// Types
import { VNode } from 'vue'
import { ClassesObject } from './../../../types'

/* @vue/component */
export default mixins(
  Routable
).extend({
  name: 'v-card',

  data: () => ({
    isMouseOver: false as boolean
  }),

  props: {
    ...VPaper.options.props,
    elevation: {
      type: [Number, String],
      default: 1
    },
    flat: Boolean,
    height: [Number, String],
    hover: [Number, String],
    /* @deprecated */
    img: String,
    maxHeight: [Number, String],
    maxWidth: [Number, String],
    /* @deprecated */
    raised: Boolean,
    tile: Boolean,
    width: [Number, String]
  },

  computed: {
    classes (): ClassesObject {
      return {
        'v-paper v-card': true,
        'v-card--tile': this.tile
      }
    },
    computedElevation (): number | string {
      if (this.raised) {
        deprecate('<v-card raised>', '<v-card elevation="3">', this)

        return 3
      }

      if (this.hover === '' && this.isMouseOver) {
        deprecate('<v-card hover>', '<v-card hover="8">', this)

        return 8
      }

      return this.flat
        ? 0
        : this.isMouseOver ? this.hover : this.elevation
    },
    listeners (): object {
      return this.hover === undefined ? {} : {
        mouseenter: () => {
          this.isMouseOver = true
        },
        mouseleave: () => {
          this.isMouseOver = false
        }
      }
    },
    styles (): object {
      const style: Record<string, any> = {
        height: convertToUnit(this.height),
        maxHeight: convertToUnit(this.maxHeight),
        maxWidth: convertToUnit(this.maxWidth),
        width: convertToUnit(this.width)
      }

      if (this.img) {
        deprecate('<v-card img="...">', 'a nested <v-img>', this)

        style.background = `url("${this.img}") center center / cover no-repeat`
      }

      return style
    }
  },

  render (h): VNode {
    const { tag, data } = this.generateRouteLink()

    data.class = {
      ...data.class,
      ...addPaperClasses({
        color: this.color,
        dark: this.dark,
        light: this.light,
        elevation: this.computedElevation
      })
    }

    data.style = this.styles
    data.on = {
      ...data.on,
      ...this.listeners
    }

    return h(tag, data, this.$slots.default)
  }
})
