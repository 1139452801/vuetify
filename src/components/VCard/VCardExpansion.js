import { VExpandTransition , VSlideYReverseTransition } from '../transitions'

export default {
  name: 'v-card-expansion',

  props: {
    show: {
      type: Boolean,
      default: false
    },
    cover: {
      type: Boolean,
      default: false
    }
  },

  render (h) {
    const children = []

    children.push(h('div', {
      'class': 'v-card__text'
    }, this.$slots.default))
    const staticClass = 'v-card__expansion ' + (this.cover ? 'v-card__expansion--cover' : 'v-card__expansion--dropdown')
    const expansionStyle = h('div', {
      'class': staticClass,
      directives: [
        {
          name: 'show',
          value: this.show
        }
      ]
    }, children)

    return h(this.cover ? VSlideYReverseTransition : VExpandTransition, [expansionStyle])
  }
}
