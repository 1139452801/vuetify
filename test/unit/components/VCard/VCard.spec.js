import { test } from '@/test'
import VCard from '@/components/VCard'

test('VCard.vue', ({ mount }) => {
  it('should render component and match snapshot', () => {
    const wrapper = mount(VCard)

    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should render card with img', () => {
    const wrapper = mount(VCard, {
      propsData: {
        img: 'image.jpg'
      }
    })

    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should render a flat card', () => {
    const wrapper = mount(VCard, {
      propsData: {
        flat: true
      }
    })

    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should render a raised card', () => {
    const wrapper = mount(VCard, {
      propsData: {
        raised: true
      }
    })

    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should render a colored card', () => {
    const wrapper = mount(VCard, {
      propsData: {
        color: 'blue lighten-1'
      }
    })

    expect(wrapper.classes()).toContain('blue')
    expect(wrapper.classes()).toContain('lighten-1')
  })

  it('should render a tile card', () => {
    const wrapper = mount(VCard, {
      propsData: {
        tile: true
      }
    })

    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should render a card with custom height', () => {
    const heightpx = '400px'
    const wrapper = mount(VCard, {
      propsData: {
        height: heightpx
      }
    })

    expect(wrapper.element.style.height).toBe(heightpx)
    expect(wrapper.html()).toMatchSnapshot()

    wrapper.setProps({
      height: 401
    })
    expect(wrapper.element.style.height).toBe('401px')
  })

  it('should render a tile card', () => {
    const wrapper = mount(VCard, {
      propsData: {
        tile: true
      }
    })

    expect(wrapper.classes()).toContain('v-card--tile')
    expect(wrapper.html()).toMatchSnapshot()
  })
})
