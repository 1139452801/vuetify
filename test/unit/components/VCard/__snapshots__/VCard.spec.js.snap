// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VCard.vue should render a card with custom height 1`] = `

<div class="v-responsive v-paper elevation-1 v-card"
     style="height: 400px;"
>
  <div class="v-responsive__content">
  </div>
</div>

`;

exports[`VCard.vue should render a flat card 1`] = `

<div class="v-responsive v-paper elevation-0 v-card">
  <div class="v-responsive__content">
  </div>
</div>

`;

exports[`VCard.vue should render a raised card 1`] = `

<div class="v-responsive v-paper elevation-3 v-card">
  <div class="v-responsive__content">
  </div>
</div>

`;

exports[`VCard.vue should render a tile card 1`] = `

<div class="v-responsive v-paper v-paper--tile elevation-1 v-card">
  <div class="v-responsive__content">
  </div>
</div>

`;

exports[`VCard.vue should render a tile card 2`] = `

<div class="v-responsive v-paper v-paper--tile elevation-1 v-card">
  <div class="v-responsive__content">
  </div>
</div>

`;

exports[`VCard.vue should render card with img 1`] = `

<div class="v-responsive v-paper elevation-1 v-card">
  <div class="v-responsive__content">
  </div>
</div>

`;

exports[`VCard.vue should render component and match snapshot 1`] = `

<div class="v-responsive v-paper elevation-1 v-card">
  <div class="v-responsive__content">
  </div>
</div>

`;
